test/EventManagerIntrospectionTrait.php

```diff
@@ -0,0 +1,56 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2015 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   http://framework.zend.com/license/new-bsd New BSD License
+ */
+
+namespace ZendTest\Mvc;
+
+use ReflectionProperty;
+use Zend\EventManager\EventManager;
+
+/**
+ * Offer methods for introspecting event manager events and listeners.
+ */
+trait EventManagerIntrospectionTrait
+{
+    public function getEventsFromEventManager(EventManager $events)
+    {
+        $r = new ReflectionProperty($events, 'events');
+        $r->setAccessible(true);
+        $listeners = $r->getValue($events);
+        return array_keys($listeners);
+    }
+
+    public function getListenersForEvent($event, EventManager $events, $withPriority = false)
+    {
+        $r = new ReflectionProperty($events, 'events');
+        $r->setAccessible(true);
+        $listeners = $r->getValue($events);
+
+        if (! isset($listeners[$event])) {
+            return $this->traverseListeners([]);
+        }
+
+        return $this->traverseListeners($listeners[$event], $withPriority);
+    }
+
+    public function traverseListeners(array $queue, $withPriority = false)
```

Comment:
It's the listeners for a specific event. In terms of the priority flag:

- false should be used if you want to potentially cast to an array. An
interesting thing with generators and iterators is that you can have keys
that are identical - but iterator_to_array will collapse those. As such,
this option does not provide keys.
- true should be used if you'll actually iterate, allowing you to get keys
and listeners.

I could collapse this method into the getListenersForEvent() method,
though; will do so shortly.
On Oct 1, 2015 7:35 AM, "Marco Pivetta" <notifications@github.com> wrote:

> In test/EventManagerIntrospectionTrait.php
> <https://github.com/zendframework/zend-mvc/pull/31#discussion_r40907593>:
>
> > +    }
> > +
> > +    public function getListenersForEvent($event, EventManager $events, $withPriority = false)
> > +    {
> > +        $r = new ReflectionProperty($events, 'events');
> > +        $r->setAccessible(true);
> > +        $listeners = $r->getValue($events);
> > +
> > +        if (! isset($listeners[$event])) {
> > +            return $this->traverseListeners([]);
> > +        }
> > +
> > +        return $this->traverseListeners($listeners[$event], $withPriority);
> > +    }
> > +
> > +    public function traverseListeners(array $queue, $withPriority = false)
>
> What does the traversal do? Does it simply allow traversing the listeners
> by their priority, emulating what the event manager does?
>
> Also, can the argument simply be the event manager (so that fetching the
> listeners becomes a detail)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zendframework/zend-mvc/pull/31/files#r40907593>.
>